document the new introduced shimv2

gnawux opened the pull request:
- [x] Manual Installation Section of the [Install Guide](https://github.com/kata-containers/documentation/blob/master/install/README.md#manual-installation), should mention configuration the CRI part
- [x] [Run Kata Containers with Kubernetes](https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#run-kata-containers-with-kubernetes) in Developer Guide
- [x] [Configure containerd to Use Kata Containers](https://github.com/kata-containers/documentation/blob/master/how-to/how-to-use-k8s-with-cri-containerd-and-kata.md#configure-containerd-to-use-kata-containers) in the kubernetes Howto
- [x] Update the [Architecture](https://github.com/kata-containers/documentation/blob/master/architecture.md) doc

Fixes: #322

gnawux commented:
CI failed due to kata-containers/tests#1018

Well, we are still working in progress, not a big deal

gnawux commented:
Here is the instruction for how to use shim v2. And I think we need a design doc for shimv2 as well and we may put it in a separate PR.

gnawux commented:
PTAL @bergwolf @lifupan @egernst

bergwolf commented on architecture.md:
```diff
@@ -638,8 +652,36 @@ a pod is **not** `Privileged` the runtime selection is done as follows:
+  [plugins.cri.containerd]
+  
+    # "plugins.cri.containerd.default_runtime" is the runtime to use in containerd.
+    [plugins.cri.containerd.default_runtime]
```
containerd cri has moved to `plugins.cri.containerd.runtimes` configuration to better match kubernetes runtime class requirements. And `untrusted_workload_runtime ` is now [deprecated](https://github.com/containerd/cri/blob/master/docs/config.md). Please use the new format and thus something like:
```toml
    [plugins.cri.containerd]
      snapshotter = "overlayfs"
      no_pivot = false
    [plugins.cri.containerd.runtimes]
      [plugins.cri.containerd.runtimes.runc]
         runtime_type = "io.containerd.runc.v1"
         [plugins.cri.containerd.runtimes.runc.options]
           NoPivotRoot = false
           NoNewKeyring = false
           ShimCgroup = ""
           IoUid = 0
           IoGid = 0
           BinaryName = "runc"
           Root = ""
           CriuPath = ""
           SystemdCgroup = false
      [plugins.cri.containerd.runtimes.kata]
         runtime_type = "io.containerd.kata.v2"
         [plugins.cri.containerd.runtimes.kata.options]
           NoPivotRoot = false
           NoNewKeyring = false
           ShimCgroup = ""
           IoUid = 0
           IoGid = 0
           BinaryName = "/usr/local/bin/containerd-shim-kata-v2"
           Root = ""
           CriuPath = ""
           SystemdCgroup = false
      [plugins.cri.containerd.runtimes.katacli]
         runtime_type = "io.containerd.runc.v1"
         [plugins.cri.containerd.runtimes.katacli.options]
           NoPivotRoot = false
           NoNewKeyring = false
           ShimCgroup = ""
           IoUid = 0
           IoGid = 0
           BinaryName = "/usr/local/bin/kata-runtime"
           Root = ""
           CriuPath = ""
           SystemdCgroup = false
```
Then users can define new runtime class resource with one of these runtime handlers: `runc`, `kata` and `kata-cli`, and reference it in their pod config, as documented in https://kubernetes.io/docs/concepts/containers/runtime-class/

The example works for runc, kata-runtime and containerd-shim-kata-v2. Please suggest better names than `kata` and `katacli` being used there though. I listed them mostly to show that we can have multiple runtime handlers within one containerd instance.

gnawux replied:
hmm, ok, will make an update for runtime class...

gnawux commented:
updated.

- I introduced a new howto dedicate for containerd and kata containers
- Make all containerd configuration related documentation link to this one
- Later we may do further re-organization of the documentation (in another issue)
  - Split different topics of architecture.md and Developer-Guide.md into small and dedicate docs
  - More illustration of how shim v2 implements CRI semantics just like kata-runtime for OCI commands
  - Move configuration of CRI-O into how-to as well, and make the kubernetes how-to support both containerd and CRI-O
  - Maybe more...

Comments are welcome @bergwolf @egernst @lifupan @jodh-intel

egernst commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+
+## Concepts
+
+### Kubernetes RuntimeClass
```
I think runtimeclass concept should be in a different document, as this is specific to k8s, not containerd

gnawux replied:
I agree `RuntimeClass` is not a containerd specific concept, but we need some background to introduce the containerd configuration.

I don't think we should have a full document on runtime class. As a result, I put a paragraph here for information.

I plan to
- move cri-o related doc contents to how-to, and update to the latest release (#333 )
- update the kubernetes howto to support both cri-o and containerd, and move the runtime class back to that document

What's your opinion, @egernst?

egernst commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+
+### Launch containers with ctr command line
+
+To run a Kata contianerd with Kata Containers through command line, you could simply run
```
Typo

gnawux replied:
Thanks, will update it.

egernst commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+           CriuPath = ""
+           SystemdCgroup = false
+      [plugins.cri.containerd.runtimes.katacli]
+         runtime_type = "io.containerd.runc.v1"
```
is this correct?

egernst replied:
I now understand that it is.  The shim naming is a bit misleading, but that's containerd's design, not ours.

jodh-intel commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+interface for [Kubernetes](https://kubernetes.io) and other CRI clients.
+
+This document is mainly written for Kata Containers v1.5.0-rc2 or above and containerd v1.2.0 or above.
+The document addressed the previous versions as well though we suggest users upgrade to the newer versions for better support.
```
s/The document addressed/The document addresses/ ?

jodh-intel commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+
+### Kubernetes RuntimeClass
+
+[RuntimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/) is a Kubernetes feature firstly 
```
s/firstly/first/

jodh-intel commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+introduced in Kubernetes 1.12 as alpha. It is the feature for selecting the container runtime configuration to 
+use to run a pod’s containers. And this feature is supported in `containerd` since [v1.2.0](https://github.com/containerd/containerd/releases/tag/v1.2.0).
+
+Before the Runtime was introduced, Kubernetes isn't aware of the difference of runtimes on the node. `kubelet`
```
- s/Runtime/RuntimeClass/ ?
- s/isn't/wasn't/

jodh-intel commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+
+Follow the instructions from the [CRI installation guide](http://github.com/containerd/cri/blob/master/docs/installation.md).
+
+And check if `containerd` is now available
```
Maybe add a colon at the end of this line?

Same comment for other sentences that introduce code blocks below.

jodh-intel commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,345 @@
+And check if `containerd` is now available
+  
+```bash
+> command -v containerd
````
Could you change the prompt to `$` for consistency with https://github.com/kata-containers/documentation/blob/master/Documentation-Requirements.md#code-blocks.

```sh

    ```bash
    $ command -v containerd
    ```
```

Same comment for commands below...

jodh-intel commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+> go get github.com/containernetworking/plugins
+> pushd $GOPATH/src/github.com/containernetworking/plugins
+> ./build_linux.sh
+> mkdir /opt/cni
```
`sudo mkdir`?

jodh-intel commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+> pushd $GOPATH/src/github.com/containernetworking/plugins
+> ./build_linux.sh
+> mkdir /opt/cni
+> cp -r bin /opt/cni/
```
`sudo cp -a`?

jodh-intel commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+> go get github.com/kubernetes-incubator/cri-tools
+> pushd $GOPATH/src/github.com/kubernetes-incubator/cri-tools
+> make
+> make install
```
`sudo -E`?

jodh-intel commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,345 @@
+To run a container with Kata Containers through containerd command line, you could simply run
+
+```bash
+> sudo ctr run --runtime io.containerd.run.kata.v2 -t --rm docker.io/library/busybox:latest hello sh
````
Is `hello` needed here?

gnawux replied:
It is the `ID` field, which is required AFAIK.

gnawux commented:
Thanks for your comments @jodh-intel, I will update it in my next morning

egernst commented:
@gnawux updates here?  I'd like to have this in for the 1.5 release if feasible (not sure if it is at this point - we're planning on Wednesday).

lifupan commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,345 @@
+           SystemdCgroup = false
+      [plugins.cri.containerd.runtimes.kata]
+         runtime_type = "io.containerd.kata.v2"
+         [plugins.cri.containerd.runtimes.kata.options]
```
By now containerd hasn't supported any kata options, so I think it's better to remove them. 
we had a PR https://github.com/containerd/containerd/pull/2916 to track on how to pass kata options from containerd to shimv2.

gnawux commented:
Thanks reviewers.

@lifupan I have updated the RuntimeClass configuration part.

@egernst I have updated the doc and add the workaround of the config file `ENV`

@jodh-intel I have updated based on your comments. Thank you very much.

jodh-intel commented:
Thanks @gnawux!

lgtm

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+The [`containerd-shim-kata-v2` (short as `shimv2` in this documentation)](https://github.com/kata-containers/runtime/tree/master/containerd-shim-v2) 
+implements the [Containerd Runtime V2 (Shim API)](https://github.com/containerd/containerd/tree/master/runtime/v2) for Kata.
+With `shimv2`, kubernetes could launch Pod and OCI compatible containers with one shim (the `shimv2`) per Pod instead
```
Lines 59-61 suggested rewrite - although I might have changed meaning because I am unclear on what the standalone kata-proxy piece refers to: 
With `shimv2`, kubernetes can launch Pod and OCI-compatible containers with one shim per Pod. Prior to `shimv2`, `2N+1` shims (i.e. a `containerd-shim` and a `kata-shim` for each container and the Pod sandbox itself) and no standalone `kata-proxy` process were used, even with vsock not available.

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+![kubernetes integration with shimv2](../arch-images/shimv2.svg)
+
+The shim v2 is introduced in containerd [v1.2.0](https://github.com/containerd/containerd/releases/tag/v1.2.0) and kata shimv2
```
Do you want "shim v2" and "kata shimv2" formatted here?

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+### Install containerd with cri plugin
+
+> **Note:** `cri` is a native plugin of containerd 1.1 and above. It is built into containerd and enabled by default. 
+> You don't need to install `cri` if you have containerd 1.1 or above. Just check if the `cri` plugin has been disabled 
```
Suggest removing the contraction / You do not need to install ...

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+Follow the instructions from the [CRI installation guide](http://github.com/containerd/cri/blob/master/docs/installation.md).
+
+And check if `containerd` is now available:
```
Suggest using "Then" / Then, check if `containerd` is now available with the following:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+### Install CNI plugins
+
+> **Note:** You don't need install CNI plugins if you don't want to use containerd with Kubernetes.
```
Suggest removing the contraction + rewrite (lines 90-91): 
You do not need to install CNI plugins if you do not want to use containerd with Kubernetes. If you have installed Kubernetes with `kubeadm`, you might have already installed the CNI plugins.

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+> **Note:** You don't need install CNI plugins if you don't want to use containerd with Kubernetes.
+> And if you have installed Kubernetes with `kubeadm`, you may have already had CNI plugins installed.
+
+You may manually install CNI plugins as follows:
```
You can manually install CNI plugins as follows:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+### Install cri-tools
+
+> **Note:** `cri-tools` is a set of tools for CRI for the development and testing. For users only want to use containerd
```
Lines 106-107 suggested rewrite: 
**Note:** `cri-tools` is a set of tools for CRI used for development and testing. Users who only want to use containerd with kubernetes can skip the `cri-tools`.

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+> **Note:** `cri-tools` is a set of tools for CRI for the development and testing. For users only want to use containerd
+> with kubernetes, you may skip the `cri-tools`.
+
+You may install the `cri-tools` from source code:
```
Suggested rewrite: 
You can install the `cri-tools` from source code as follows:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+### Configure containerd to use Kata Containers
+
+By default, the configuration of containerd locates at `/etc/containerd/config.toml`, and the 
```
locates / located:
By default, the configuration of containerd is located at `/etc/containerd/config.toml`, and the

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+      conf_dir = "/etc/cni/net.d"
+```
+
+Let's add Kata Containers to the configurations.
````
Is this referring to the upcoming section? If yes, suggested rewrite: 
The following section outlines how to add Kata Containers to the configurations.

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+The  RuntimeClass is suggested.
+
+The following configuration include three runtime classes:
```
include / includes 
The following configuration includes three runtime classes:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+The  RuntimeClass is suggested.
+
+The following configuration include three runtime classes:
+- `plugins.cri.containerd.runtimes.runc`: the runC, and it is the default runtime;
```
Suggest using periods to punctuate each bullet rather than a semi-colon

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+The following configuration include three runtime classes:
+- `plugins.cri.containerd.runtimes.runc`: the runC, and it is the default runtime;
+- `plugins.cri.containerd.runtimes.kata`: it is magic in containerd (ref [the doc here](https://github.com/containerd/containerd/tree/master/runtime/v2#binary-naming)), in which the dot connected string `io.containerd.kata.v2` will be translated to `containerd-shim-kata-v2`, i.e. the binary name of the kata implementation of shim v2 API;
```
Lines 151-152 suggested rewrite (I very easily could have totally changed meaning here - hopefully not): 
- `plugins.cri.containerd.runtimes.kata`: The function in containerd (reference [the document here](https://github.com/containerd/containerd/tree/master/runtime/v2#binary-naming)) where the dot-connected string `io.containerd.kata.v2` is translated to `containerd-shim-kata-v2` (i.e. the binary name of the kata implementation of shim v2 API), [Containerd Runtime V2 (Shim API)](https://github.com/containerd/containerd/tree/master/runtime/v2).

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+- `plugins.cri.containerd.runtimes.runc`: the runC, and it is the default runtime;
+- `plugins.cri.containerd.runtimes.kata`: it is magic in containerd (ref [the doc here](https://github.com/containerd/containerd/tree/master/runtime/v2#binary-naming)), in which the dot connected string `io.containerd.kata.v2` will be translated to `containerd-shim-kata-v2`, i.e. the binary name of the kata implementation of shim v2 API;
+[Containerd Runtime V2 (Shim API)](https://github.com/containerd/containerd/tree/master/runtime/v2), the 
+- `plugins.cri.containerd.runtimes.katacli`: the `containerd-shim-runc-v1` will call `kata-runtime`, which is the legacy way deprecated by the `containerd-shim-kata-v2`:
```
Suggested rewrite: 
- `plugins.cri.containerd.runtimes.katacli`: the `containerd-shim-runc-v1` calls `kata-runtime`, which is the legacy process

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+Based on the discussion in [containerd/containerd#2916](https://github.com/containerd/containerd/pull/2916),
+the containerd configuration will have a field for runtime configuration files (likely to be`config_file`). However, 
+the field has not been available yet and the `shimv2` will search configuration file in the default paths 
```
Lines 189-191 suggested rewrite: 
the field is not available yet, and the `shimv2` will search the configuration file in the default paths, similar to `kata-runtime` cli): `/etc/kata-containers/configuration.toml`, `/usr/share/defaults/kata-containers/configuration.toml`. We will update this document once the `config_file` field is available.

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+(same as `kata-runtime` cli): `/etc/kata-containers/configuration.toml`, `/usr/share/defaults/kata-containers/configuration.toml`. 
+And the document will be updated once the `config_file` field is available.
+
+If you want to specify a configure file now, here is a workaround. The `shimv2` accept an environment variable,
```
Lines 193-194 suggested rewrite: 
If you want to specify a configure file now, the following provides a workaround. The `shimv2` accepts an environment variable, `KATA_CONF_FILE` for the config file path. Then, you can create a shell script with the following:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+#### Kata Containers as the runtime for untrusted workload
+
+For the without RuntimeClass support cases, we could use the legacy annotation way to support using Kata Containers
```
Lines 210-212 suggested rewrite: 
For cases without RuntimeClass support, we can use the legacy annotation method to support using Kata Containers for an untrusted workload. With the following configuration, you can run trusted workloads with a runtime such as `runc` and then, run an untrusted workload with Kata Containers:

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+      runtime_type = "io.containerd.kata.v2"
+```
+
+For the earlier versions of Kata Containers and containerd that don't support Runtime V2 (Shim API), you could use the alternative configuration:
````
Suggested rewrite: 
For the earlier versions of Kata Containers and containerd that do not support Runtime V2 (Shim API), you can use the following alternative configuration:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+#### Kata Containers as the default runtime
+
+If you want to set Kata Containers as the only runtime in the deployment, you could simply configure as follows:
```
could / can
If you want to set Kata Containers as the only runtime in the deployment, you can configure as follows:

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+      runtime_type = "io.containerd.kata.v2"
+```
+
+Or for the earlier versions of Kata Containers and containerd that don't support Runtime V2 (Shim API), you could use the alternative configuration:
````
Suggested rewrite: 
Alternatively, for the earlier versions of Kata Containers and containerd that do not support Runtime V2 (Shim API), you can use the following alternative configuration:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+### Configuration for cri-tools
+
+> **Note:** If you have skipped the [Install cri-tools](#install-cri-tools) section, you should probably skip this section too.
```
Suggested rewrite: 
**Note:** If you skipped the [Install cri-tools](#install-cri-tools) section, you can skip this section too.

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+> **Note:** If you have skipped the [Install cri-tools](#install-cri-tools) section, you should probably skip this section too.
+
+Firstly, add the CNI configuration in the containerd configuration. 
```
Firstly / First
First, add the ...

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+Firstly, add the CNI configuration in the containerd configuration. 
+
+Here is the configuration if you installed CNI as the *[Install CNI plugins](#install-cni-plugins)* section.
```
The following is the configuration if you installed CNI as the *[Install CNI plugins](#install-cni-plugins)* section outlined.

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+}
+```
+
+And reference the configuration directory through containerd `config.toml`:
````
Next, reference the configuration directory through containerd `config.toml`:

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+### Launch containers with ctr command line
+
+To run a container with Kata Containers through containerd command line, you could simply run:
```
Suggested rewrite: 
To run a container with Kata Containers through the containerd command line, you can run the following:

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+$ sudo ctr run --runtime io.containerd.run.kata.v2 -t --rm docker.io/library/busybox:latest hello sh
+```
+
+which will launch a busybox container named as `hello` and it will be removed by `--rm` after it quits.
````
Suggested rewrite: 
This launches a busybox container named `hello` that is removed by `--rm` after it quits.

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+### Launch Pods with crictl command line
+
+With `crictl` command line of `cri-tools`, you may specify runtime class with `-r` or `--runtime` flag,
```
Lines 327-328 suggested rewrite: 
With the `crictl` command line of `cri-tools`, you can specify runtime class with `-r` or `--runtime` flag. Use the following to launch Pod with `kata` runtime class with the pod in [the example](https://github.com/kubernetes-sigs/cri-tools/tree/master/docs/examples)

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+36e23521e8f89fabd9044924c9aeb34890c60e85e1748e8daca7e2e673f8653e
+```
+
+And you may add container to the launched Pod:
````
You can add container to the launched Pod with the following:

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+1aab7585530e62c446734f12f6899f095ce53422dafcf5a80055ba11b95f2da7
+```
+
+Then start it:
````
Now, start it with the following:

klynnrif commented on how-to/containerd-kata.md:
````diff
@@ -0,0 +1,357 @@
+1aab7585530e6
+```
+
+In Kubernetes, you need create RuntimeClass resource and add RuntimeClass field in the Pod Spec (see the [document](https://kubernetes.io/docs/concepts/containers/runtime-class/)).
````
Suggested rewrite: 
In Kubernetes, you need to create a RuntimeClass resource and add the RuntimeClass field in the Pod Spec (see this [document](https://kubernetes.io/docs/concepts/containers/runtime-class/) for more information).

klynnrif commented on how-to/containerd-kata.md:
```diff
@@ -0,0 +1,357 @@
+
+In Kubernetes, you need create RuntimeClass resource and add RuntimeClass field in the Pod Spec (see the [document](https://kubernetes.io/docs/concepts/containers/runtime-class/)).
+
+And if RuntimeClass is not supported, you may have to use the following annotation in a Kubernetes pod to identify as an untrusted workload:
```
Suggested rewrite: 
If RuntimeClass is not supported, you can use the following annotation in a Kubernetes pod to identify as an untrusted workload:

klynnrif commented on how-to/how-to-use-k8s-with-cri-containerd-and-kata.md:
```diff
@@ -20,33 +18,28 @@ The Kubernetes cluster will use the
 [Kata Containers](https://katacontainers.io) to launch untrusted workloads.
 
+For Kata Containers 1.5.0-rc2 and above, we will use [`containerd-shim-kata-v2` (short as `shimv2` in this documentation)](../design/shimv2.md) 
+launch Kata Containers. While for the previous version of Kata Containers, the Pods will be launched with `kata-runtime`.
```
Suggested rewrite: 
to launch Kata Containers. For the previous version of Kata Containers, the Pods are launched with `kata-runtime`.

klynnrif commented on how-to/how-to-use-k8s-with-cri-containerd-and-kata.md:
```diff
@@ -20,33 +18,28 @@ The Kubernetes cluster will use the
+## Install and configure containerd
 
-## Install Kata Containers
+Firstly, follow the [How to use Kata Containers and Containerd](containerd-kata.md) to install
```
First, follow the [How to use Kata Containers and Containerd](containerd-kata.md) to install and configure containerd. Then, make sure the containerd works with the [examples in it](containerd-kata.md#run).

gnawux commented:
Thank you @klynnrif, will check them in hours.

gnawux commented:
updated.

Thank you very much, @klynnrif!
